File: Java/LifecycleDemo/app/src/main/java/com/example/lifecycledemo/MainActivity.java
package com.example.lifecycledemo;

import android.annotation.SuppressLint;
import android.os.Bundle;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import android.view.View;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private TextView countText;

    MainActivityViewModel mainActivityViewModel;

    @SuppressLint("SetTextI18n")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        countText = findViewById(R.id.count_text);

        mainActivityViewModel = ViewModelProviders.of(this).get(MainActivityViewModel.class);
        LiveData<Integer> count = mainActivityViewModel.getInitialCount();

        count.observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(Integer integer) {
                countText.setText("Count is: "+ integer);
            }
        });

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mainActivityViewModel.updateCurrentCount();
            }
        });
    }
}

File: Kotlin/CoroutineDemo/app/src/main/java/com/example/coroutinedemo/MainActivity.kt
package com.example.coroutinedemo

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class MainActivity : AppCompatActivity() {
    private var count = 0
    private lateinit var mainActivityViewModel : MainActivityViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        View Model scope
        mainActivityViewModel = ViewModelProvider(this).get(MainActivityViewModel::class.java)

//        commented because of using live data
//        mainActivityViewModel.getUserData()
        mainActivityViewModel.users.observe(this, Observer {myUsers->
            myUsers.forEach {
                Log.i("MyTag"," name is ${it.name}")
            }

        })
//        View Model scope

        btnCount.setOnClickListener {
            tvCount.text = count++.toString()
        }
        btnDownloadUserData.setOnClickListener {
            CoroutineScope(Dispatchers.IO).launch {
                downloadUserData()
            }
        }
    }

//    view on UI
    private suspend fun downloadUserData() {
        for (i in 1..50000) {
            withContext(Dispatchers.Main) {
                tvUserMessage.text = "Downloading user $i in ${Thread.currentThread().name}"
            }
        }
    }

//    using in background
//    private fun downloadUserData() {
//        for (i in 1..50000) {
//            Log.i("MyTag", "Downloading user $i in ${Thread.currentThread().name}")
//        }
//    }
}
